Provide capability to use UTC time notation for console logs

tomkerkhove opened the pull request:
Provide capability to use UTC time notation for console logs via `ConsoleLoggerOptions.UseUtcTimezone`.

This is turned off by default to be backwards compatible.

Addresses #1808

In terms of testing I would suggest to provide a similar test to `WriteCore_LogsCorrectTimestamp` which uses UTC option and verifies stamp or a similar one - Thoughts?

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/Extensions?pullRequest=1896) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/Extensions?pullRequest=1896) before we can accept your contribution.<br/><br/>:x: tomkerkhove [sign now](https://cla.dotnetfoundation.org/aspnet/Extensions?pullRequest=1896)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/Extensions?pullRequest=1896) it.</sub>

gfoidl commented on src/Logging/Logging.Console/src/ConsoleLoggerOptions.cs:
```diff
@@ -37,5 +37,10 @@ public ConsoleLoggerFormat Format
+        /// <summary>
+        /// Gets or sets indication whether or not UTC timezone should be used to for timestamps in logging messages. Defaults to <c>false</c>.
+        /// </summary>
+        public bool UseUtcTimezone { get; set; } = false;
```
```suggestion
        public bool UseUtcTimezone { get; set; }
```
`false` is already the default.

tomkerkhove replied:
True but I wanted to be more explicit - However it doesn't really make sense so I'll remove it!

gfoidl commented on src/Logging/Logging.Console/src/ConsoleLogger.cs:
```diff
@@ -137,11 +137,16 @@ private LogMessageEntry CreateDefaultLogMessage(StringBuilder logBuilder, LogLev
             var timestampFormat = Options.TimestampFormat;
+            if (timestampFormat != null)
+            {
+                timestamp = Options.UseUtcTimezone ? DateTime.UtcNow.ToString(timestampFormat) : DateTime.Now.ToString(timestampFormat);
```
```suggestion
                var dateTime = Options.UseUtcTimezone ? DateTime.UtcNow : DateTime.Now;
                timestamp = dateTime.ToString(timestampFormat);
```
But I think this is more a matter of taste.

tomkerkhove replied:
Fine for me, will send update asap

tomkerkhove commented:
@gfoidl I've incorporated your suggested and added a test similar to the existing one.

However, they do not verify the output so I'm curious how you would approach it?
Should I just deserialize them and verify the timezone? Open to suggestions.

gfoidl commented on src/Logging/test/ConsoleLoggerTest.cs:
```diff
@@ -382,14 +382,50 @@ public void WriteCore_LogsCorrectTimestamp(ConsoleLoggerFormat format, LogLevel
+            // Act
+            logger.Log(level, 0, _state, ex, _defaultFormatter);
+
+            // Assert
```
Hm, there is no assert that UTC is written, so this test is more or less a duplication of the already existing test for normal datetime. (just see you commented on that too :wink:).

> Should I just deserialize them and verify the timezone?

That's possible (but then you have to assert / take into account that current timezone != UTC).

Maybe it's better to (at least in my projects I would do) extract the creation of the `timestamp` in `ConsoleLogger` to a separate method that is testable?

Something like 
```c#
private static string GetTimestamp(string timestampFormat, bool useUtcTimezone)
    => GetTimestamp(timestampFormat, useUtcTimezone, DateTime.Now);

internal static string GetTimestamp(string timestampFormat, bool useUtcTimezone, DateTime now)
{
    if (timestampFormat == null)
        return null;

    if (useUtcTimezone)
        now = now.ToUniversalTime();

    return now.ToString(timestampFormat);
}
```
Then the second method is testable.

tomkerkhove replied:
That's one approach but who says that method is being called? We have no guarantee about that actually so not sure if that's going to cover everything.

gfoidl replied:
You're right. So your approach is a better end-to-end test.

tomkerkhove replied:
So what I'd propose then would be:
1. Change format to `yyyyMMddHHmmsszzz`
2. Get tracked timestamp via regex similar to https://github.com/aspnet/Extensions/pull/1896/files#diff-a27bf784d2879ac5f61a0d1a8573fc80R418
3. Parse to DateTime and check timezone

Thoughts?

I'm not a regex expert so help would be ideal on that 😅

tomkerkhove replied:
This should be there, except for explicitly setting the culture to a non-UTC compatible one for the non-UTC test.

I tried to use the folling but it didn't work:
```csharp
var cultureInfo = new CultureInfo("en-US");
CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;
```

tomkerkhove commented:
Looks like my PR is messed up due to the rebase, going to start with a clean branch.

tomkerkhove commented:
Closing in favor of #2420 with a clean branch.
